### PR TITLE
Add activation quantizers for QAT

### PR DESCRIPTION
**Pull request by elad-c**
* Refactor KerasModelBuilder so it's not an abstact class. Remove QAT model builder
* Fix input quantization wrapping: Add Identity layer after input layers to be wrapped with an activation quantization wrapper. Input layer in back2framework renamed
* Fix exporter trasnform pattern to match new Inout layer name
* QAT Finalize disabled. Awaiting integration with Exporter

### CODE REVIEW

File: model_compression_toolkit/qunatizers_infrastructure/keras/quantize_wrapper.py [hunk 1]
```diff
@@ -172,11 +179,28 @@ def call(self, inputs, training=None, **kwargs):
...
             self.set_quantize_weights(quantized_weights)
 
-            args = tf_inspect.getfullargspec(self.layer.call).args
-            if TRAINING in args:
+            args_spec = tf_inspect.getfullargspec(self.layer.call).args
+            if TRAINING in args_spec:
                 outputs = self.layer.call(inputs, training=training, **kwargs)
             else:
                 outputs = self.layer.call(inputs, **kwargs)
+            if self.dispatcher.num_act_quantizers > 0:
```
Comment (eladc-git) [inline model_compression_toolkit/qunatizers_infrastructure/keras/quantize_wrapper.py:187]: maybe add space and comment "Quantize all activations" so it is clear that this block deals with activations

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py [hunk 2]
```diff
@@ -143,3 +144,97 @@ def __call__(self,
...
+        Args:
+            tensor_shape: Tensor shape the quantizer quantize.
+            name: Prefix of variables names.
+            layer: Layer to add the variables to. The variables are saved
+            in the layer's scope.
+
+        Returns:
+            Dictionary of new variables.
+        """
+        ptq_threshold_tensor = layer.add_weight(
```
Comment (eladc-git) [inline model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:196]: where do you use it?

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py [hunk 3]
```diff
@@ -143,3 +144,97 @@ def __call__(self,
...
+            shape=(),
+            initializer=tf.keras.initializers.Constant(1.0),
+            trainable=False)
+        ptq_threshold_tensor.assign(self.np_threshold_values)
+
+        fq_min = layer.add_weight(
+            name + FQ_MIN,
+            shape=(),
+            initializer=tf.keras.initializers.Constant(-1.0),
+            trainable=False)
```
Comment (eladc-git) [inline model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:207]: Why False? you want to train it in QAT don't you?

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py [hunk 4]
```diff
@@ -143,3 +144,97 @@ def __call__(self,
...
+            inputs: Input tensor to quantize.
+            training: Whether the graph is in training mode.
+
+        Returns:
+            The quantized tensor.
+        """
+
+        _min = self.quantizer_parameters[FQ_MIN]
+        _max = self.quantizer_parameters[FQ_MAX]
+        q_tensor = tf.quantization.fake_quant_with_min_max_vars(inputs, _min, _max,
```
Comment (eladc-git) [inline model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:237]: Where are the constraints for symmetric or PO2 ? Am I missing something?

**Review by eladc-git — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py (hunk 2 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:196]: will be refactored next PR

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py (hunk 3 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:207]: currently used as fixed quantizer only

File: model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py (hunk 4 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/ste_rounding/symmetirc_ste.py:237]: will be added next PR

**Review by eladc-git — APPROVED** [review]

File: model_compression_toolkit/qat/keras/qat_model_builder.py [hunk 5]
```diff
@@ -45,7 +45,9 @@ def _is_qat_applicable(node: common.BaseNode,
         A boolean whether the layer is to be wrapped with a QuantizeWrapper
     """
 
-    return fw_info.is_kernel_op(node.type) and node.is_weights_quantization_enabled()
+    is_weight_quantization = fw_info.is_kernel_op(node.type) and node.is_weights_quantization_enabled()
```
Comment (reuvenperetz) [inline model_compression_toolkit/qat/keras/qat_model_builder.py:48]: I don't think we need fw_info here as it takes the quantization info from the node (which comes from the TPC)

File: model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py [hunk 6]
```diff
@@ -18,26 +18,34 @@
...
 
 def quantization_dispatcher_builder(n: common.BaseNode,
                                     fw_info: FrameworkInfo,
-                                    method2quantizer: Dict[
-                                        qi.QuantizationMethod, qi.BaseKerasQuantizer] = METHOD2QUANTIZER) -> qi.KerasNodeQuantizationDispatcher:
+                                    method2weightquantizer: Dict[
+                                        qi.QuantizationMethod, qi.BaseKerasQuantizer] = METHOD2WEIGHTQUANTIZER,
+                                    method2actquantizer: Dict[
+                                        qi.QuantizationMethod, qi.BaseKerasQuantizer] = METHOD2ACTQUANTIZER
+                                    ) -> qi.KerasNodeQuantizationDispatcher:
```
Comment (reuvenperetz) [inline model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py:39]: Maybe it's better to use None in the signature (and use default dicts if they are None)?

File: model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py [hunk 7]
```diff
@@ -47,10 +55,16 @@ def quantization_dispatcher_builder(n: common.BaseNode,
...
             if qunatizer_class is None:
                 common.Logger.error(
                     f'Unknown Quantiztion method: {n.final_weights_quantization_cfg.weights_quantization_method}')
             nqd.add_weight_quantizer(attr, qunatizer_class(n.final_weights_quantization_cfg))
 
+    if n.is_activation_quantization_enabled():
+        # single output -> normalize to list of output_shapes
+        output_shapes = n.output_shape if isinstance(n.output_shape[0], (list, tuple)) else [n.output_shape]
+        qunatizer_class = method2actquantizer.get(n.final_activation_quantization_cfg.activation_quantization_method)
+        nqd.activation_quantizers = [qunatizer_class(n.final_activation_quantization_cfg)] * len(output_shapes)
```
Comment (reuvenperetz) [inline model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py:68]: Is it correct for reused layers? What if its outputs should be quantized differently?

File: tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py [hunk 8]
```diff
@@ -28,25 +28,30 @@
...
 class QuantizationAwareTrainingTest(BaseKerasFeatureNetworkTest):
     def __init__(self, unit_test, layer, weight_bits=2, activation_bits=4, finalize=False,
-                 weights_quantization_method=mct.target_platform.QuantizationMethod.POWER_OF_TWO, test_loading=False):
+                 weights_quantization_method=mct.target_platform.QuantizationMethod.POWER_OF_TWO,
+                 activation_quantization_method=mct.target_platform.QuantizationMethod.POWER_OF_TWO,
+                 test_loading=False):
         self.layer = layer
         self.weight_bits = weight_bits
         self.activation_bits = activation_bits
+        assert finalize is False, 'MCT QAT Finalize is disabled until exporter is fully supported'
```
Comment (reuvenperetz) [inline tests/keras_tests/feature_networks_tests/feature_networks/qat/qat_test.py:37]: Let's talk about it

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/keras/back2framework/instance_builder.py [hunk 9]
```diff
@@ -86,10 +73,15 @@ def node_builder(n: common.BaseNode) -> Layer:
...
     framework_attr = copy.copy(n.framework_attr)
+    if n.layer_class is InputLayer:
+        # replace input node with identity, so can wrap it with QuantizationWrapper
+        _layer_class = Layer  # Identity
+        framework_attr = {}
+    else:
+        _layer_class = n.layer_class
     framework_attr[LAYER_NAME] = n.name  # Overwrite framework name to identical graph node name
-    node_instance = n.layer_class.from_config(framework_attr)  # Build layer from node's configuration.
+    node_instance = _layer_class.from_config(framework_attr)  # Build layer from node's configuration.
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/keras/back2framework/instance_builder.py:84]: Not in the instance builder. It's separatelly before that in the keras model builder (line 152)

File: model_compression_toolkit/core/keras/back2framework/keras_model_builder.py [hunk 10]
```diff
@@ -109,9 +109,9 @@ def __init__(self,
                          return_float_outputs)
 
         # Build an OperationHandler to handle conversions from graph nodes to Keras operators.
-        self.oh = OperationHandler(self.graph, wrapper)
+        self.oh = OperationHandler(self.graph)
+        self.wrapper = wrapper
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/keras/back2framework/keras_model_builder.py:113]: No, it's removed completely from the OperationHandler and handled in the model builder instance builder

**Review by reuvenperetz — APPROVED** [review]

File: model_compression_toolkit/core/keras/back2framework/keras_model_builder.py [hunk 11]
```diff
@@ -246,19 +257,20 @@ def _run_operation(self,
...
             layer to the input tensors.
         """
-
         if len(input_tensors) == 0:  # Placeholder handling
             out_tensors_of_n_float = input_nodes_to_input_tensors[n]
-            out_tensors_of_n = out_tensors_of_n_float
-            if n.is_activation_quantization_enabled():
+            if self.wrapper is not None:
+                # if a wrapper is defined, add an identity layer for cloning
+                out_tensors_of_n = op_func(out_tensors_of_n_float)
```
[the remark this answers is not in the document]
Comment (elad-c) [reply model_compression_toolkit/core/keras/back2framework/keras_model_builder.py:269]: op_func for Input is Identity (defined in node_builder) which will be wrapped with activation quantizer later in the build_model

File: model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py (hunk 7 above)
Comment (elad-c) [reply model_compression_toolkit/qat/keras/quantizer/quantization_dispatcher_builder.py:68]: Even for a reused layers, the quantization must be the same.

**Pull request merged**